### PR TITLE
Try: Fixed Mobile Toolbar, Yet Again

### LINKED ISSUE #14419: Block toolbar DOM order and visual order mismatch in the responsive view
Splitting this out from #9739 

In the responsive view, the block toolbar changes position and it's placed _after_ the block together with the block "movers" and the Add block button:

Screenshot 2019-03-13 at 18 39 17

However, to my understanding there's some DOM repositioning but it's mainly a visual repositioning. The DOM order is mostly the same as in the desktop view. This way, visual order and DOM order mismatch and don't make any sense.

Worth reminding [many screen reader users use a keyboard when using their mobile devices](https://webaim.org/projects/screenreadersurvey7/#mobilekeyboard) so keyboard operability is relevant also in the mobile view.

It's also a WCAG failure but, regardless, it makes operating with a keyboard nearly impossible and should be fixed. If a reordering is desired, it should change also the DOM order to match the visual order.

### LINKED ISSUE #18632: Explore fixed top toolbar on mobile (again)
## Context

On the navigation block, we ran into an [issue](https://github.com/WordPress/gutenberg/issues/18401#issuecomment-553935644) with the way the block toolbar behaves on mobile. @mtias suggested that rather than looking at this issue in isolation, it might be worth exploring the topic of having a fixed top toolbar on mobile again.

I'm opening this issue so we can track and discuss any findings in a central location. It seemed like several people might be thinking about this topic at the moment.

## Prior exploration

Fixing the toolbar to the top on mobile was an original design intent. Unfortunately, due to issues with Safari on iOS, this was deemed not feasible at the time.

@jasmussen summarised the issues nicely in [this comment](https://github.com/WordPress/gutenberg/issues/7479#issuecomment-410988762):

> On Mobile Safari, when the soft keyboard shows and/or when the text caret is in text, **the viewport moves upwards but is not resized**. Additionally, **the scroll bounce effect messes with fixed and sticky positionings**, meaning that we can't rely on any `fixed` or `sticky` element to retain its intended position.

43766357-7fca8240-9a32-11e8-87fa-3df589b6e112

Further related tickets and discussions I came across while digging into this:

https://github.com/WordPress/gutenberg/pull/6307#issuecomment-383495889, https://github.com/WordPress/gutenberg/pull/6347#issuecomment-385473537, https://github.com/WordPress/gutenberg/issues/7479#issuecomment-410988762

## What changed since the last time we looked at this?

Most notably iOS 13 released and with it some changes to how mobile Safari behaves.

https://developer.apple.com/documentation/safari_release_notes/safari_13_release_notes

Previously when using the soft-keyboard, instead of reducing the height of the viewport, Safari on iOS would simply push the viewport upwards. The element while still technically being correctly fixed to the top of the viewport was pushed out of view by the soft-keyboard. 

This issue is now fixed as @jasmussen showed in this example:

ezgif com-resize

Also, in https://github.com/WordPress/gutenberg/pull/18044 @youknowriad is working on refactoring the layout component to separate the UI from the content which according to [this comment](https://github.com/WordPress/gutenberg/pull/18624#discussion_r348403708) will hopefully make this change simpler.

## What hasn't changed since the last time we looked at this?

One of the issue(s?) that remains is that it is still possible to scroll beyond the end of the viewport. This is caused by the bouncy overscroll behavior we experience on mobile Safari when scrolling beyond the bottom or top of a page. As visible in the example above, this causes whitespace to appear where it shouldn't.

This is still something we would need to solve. 

There is a CSS WG specification called [CSS Overscroll Behavior Module Level 1](https://drafts.csswg.org/css-overscroll/) that is designed to address this issue in the future. There's a pretty neat explanation on how this will work in [this post](https://developer.mozilla.org/en-US/docs/Web/CSS/overscroll-behavior) on MDN.

Unfortunately, this feature is [not supported](https://caniuse.com/#feat=mdn-css_properties_overscroll-behavior) on mobile Safari at this point. It is however actively being worked on, I found [this related ticket](https://bugs.webkit.org/show_bug.cgi?id=176454#c28) on the official bug tracker. In the meantime, if we want to progress on this issue, we'd have to find a different solution to the issue though.

### PR DESCRIPTION
**Pull request by jasmussen**
closes #18632
closes #14419

This branch is forked off of #18663 and takes it slightly further, but it touches some sensitive sections of code along the way so therefore separate. 

The good news: as is, this branch can enable the following Mobile Safari experience: 
good

The bad news is, the following issue is still present:

bad

And what is that issue exactly, you ask? It is explained in more depth in #18632, but it boils down to this:

**We have two scrolling containers, and we can't prevent the 1st one from scrolling.**

We _only_ want `.edit-post-editor-regions` to scroll. When you only scroll that container, things are good. However, `html` also scrolls, and we can't prevent it from scrolling. This is what's causing the top toolbar to sometimes disappear off screen, or what makes a white area at the bottom appear. The `html` element scrolls when:

- The soft-keyboard first appears, Mobile Safari is "helpful" to scroll your content into view. That means the first time you open the soft keyboard, you have to swipe down a bit to see the top toolbar.
- When you scroll to the bottom of `.edit-post-editor-regions`, pause, and scroll further.
- When you start a scroll by pressing the top toolbar.

As far as I can tell, there are two ways to solve that scroll issue:

1. Wait for the `overscroll-behavior` feature to land in Mobile Safari ([read more](https://developer.mozilla.org/en-US/docs/Web/CSS/overscroll-behavior))
2. Dive into JavaScript hacks.

There is an argument to be made that the currently shipping mobile web experience on iOS is *so horrendously bad* that the behavior in this branch, as is, is an improvement. For that reason, we might want to consider shipping this as is, and hope for Mobile Safari to improve soon. I'm increasingly skeptical that diving into JS hacks is a good idea, and even if we do, we could potentially do it separately from this PR. What do you think?

If you all agree this PR is a step forward, regardless of the Mobile Safari bugs, there are a couple of things that need to happen in order for this PR to be shippable:

- We need to carve out a permanent space for the top toolbar on mobile, as opposed to showing it only when a block is selected. Otherwise it covers content at the top of the viewport. 
- This needs to be fairly widely tested. While it's solid for me in Chrome, Safari, Firefox, and the iOS Simulator, it would be good to test on a physical iPhone because there was some scroll behavior that seemed to "lock up" for me at times, and I'm not sure if that was the simulator or not. It should also be tested on Android, and in IE for desktops.

### CODE REVIEW

**Comment by jasmussen** [comment]
Depending on discussion, this PR fixes #18632 and replaces/closes #18663.

File: packages/edit-post/src/components/editor-regions/style.scss [hunk 1]
```diff
@@ -1,47 +1,74 @@
...
+	right: 0;
+	bottom: 0;
 
-		// On Desktop position the container as fixed to avoid scroll bleed.
-		position: fixed;
-		top: $admin-bar-height;
-		left: 0;
-		right: 0;
-		bottom: 0;
+	// Cancel the fixed positioning used on Mobile.
```
Comment (youknowriad) [inline packages/edit-post/src/components/editor-regions/style.scss]: The comment is a bit misleading as we're not canceling the fixed positioning but more adjustinig the "top" to show the WPAdmin header right?

File: packages/edit-post/src/components/editor-regions/style.scss [hunk 2]
```diff
@@ -1,47 +1,74 @@
...
+	// See also: https://drafts.csswg.org/css-overscroll/
+	overscroll-behavior-y: none;
 }
 
 .edit-post-editor-regions__content {
 	flex-grow: 1;
-
-	// On Mobile the header is fixed to keep HTML as scrollable.
-	@include break-medium() {
-		overflow: auto;
```
Comment (youknowriad) [inline packages/edit-post/src/components/editor-regions/style.scss]: Is this change needed? I mean in desktop the content area should still be scrollable right? (I didn't test yet)

Comment (youknowriad) [reply packages/edit-post/src/components/editor-regions/style.scss]: I think this probably broke the sidebar scrolling. It is scrolling with the main area now, it shouldn't.

**Comment by jasmussen** [comment]
Great catches, Riad. Addressed.

**Review by youknowriad — APPROVED** [review]
I moved the mobile native files to their new adequate location.

LGTM 👍

**Comment by jasmussen** [comment]
Thank you for the approval! I take that as a sign that you agree that the trade-off is worth it, given that it's better than what we have today! 

I still need to do some additional testing, including on some physical devices, and would appreciate anyone who has the time to help with that. @frontdevde do you have bandwidth?

**Comment by youknowriad** [comment]
Should we land this in Gutenberg 7.0 to allow it to be tested by plugin users?

**Comment by jasmussen** [comment]
Yes, I think so. Let me test it on my Android simulator in a bit and report back. I am expecting it to just work, because it runs Chrome and it should be the same as the web inspector on desktop, but I'd like to just verify.

**Comment by jasmussen** [comment]
This branch MIGHT need a little more time in the oven. I just realized that if you visit the post editor URL directly, there isn't a way to get to the rest of WPadmin. This is admittedly edgecase, but we might need at least the "back button". Which is easy to unhide, but takes up space, so needs testing with translations. OR it needs work to unhide the adminbar again.

**Comment by jasmussen** [comment]
Actually @frontdevde helped me uncover an additional issue where the toolbar is simply out of view. I'm going to investigate what makes that happen. But it means we do need more time for this one unfortunately.

**Comment by frontdevde** [comment]
I've tested the current version of the branch on a physical device, iPhone X running iOS 13.1.3. It looks and works nicely until the soft-keyboard comes into the picture. Then, unfortunately, the behavior becomes quite inconsistent. Sometimes the block toolbar gets pushed up just by a few pixels, other times it sticks all the way at to the top of the page.

ezgif com-video-to-gif (1)

(Had to reduce the frame rate to stay within the 10mb GitHub file upload limit.)

**Comment by frontdevde** [comment]
I just updated the physical device to 13.2.3 just to make sure the version doesn't skew our testing. Unfortunately, after updating the result remains the same as seen in the video above.

**Comment by jasmussen** [comment]
So, I've managed to narrow down what the issue with, and there's good news and bad news. Watch the GIF carefully:

simulator

What happens here is:

- if you click an input field close to the top of the viewport, the `html` element is not scrolled very far upwards at all. 
- if you click an input field that's near the bottom of the viewport, the `html` element is scrolled VERY far upwardd

So what happens here is the Mobile Safari tries to be helpful and scrolls the viewport to vertically center that which has focus, and it assumes that the `html` element is the one to scroll because _why wouldn't it be_. 

The good news:

- We may be able to fix the scroll issue by triggering a JavaScript scroll on mobile when a block is selected, a la https://stackoverflow.com/a/57774220. this would have to scroll the `html` element, and it would be able to scroll it into perfect un-cropped view
- We can potentially simplify the CSS compared to what I wrote in this branch, bring back some of the sticky magic.

The bad news:

- This needs a little JavaScript, and will not be shippable without it.

**Comment by frontdevde** [comment]
As outlined by @jasmussen above, this specific issue seems to be caused by the browser's scrolling behavior that occurs when we set the focus on a text block input.

This sounded like a use case for `preventScroll` on focus. The Focus management APIs [spec](https://html.spec.whatwg.org/multipage/interaction.html#focus-management-apis) reads

> By default, this method also scrolls the element into view. Providing the preventScroll option and setting it to true prevents this behavior.

Unfortunately, currently the `preventScroll` option, while available in Chrome and Firefox, is unsurprisingly [not supported](https://caniuse.com/#feat=mdn-api_htmlelement_focus_preventscroll_option) in Safari, yet.

**Comment by frontdevde** [comment]
@jasmussen and I have continued to look into this. I was able to confirm Joen's hypothesis as to what is causing the issue of the disappearing toolbar. Here's a recording of one of the fixes I tried:

ezgif com-video-to-gif (2)

This already works quite well. We have two options: a) find a way to prevent the browser from attempting to scroll at all (what the unfortunately unsupported `preventScroll` was designed for or b) 
scroll the frame back into place right away (which gives us a result similar to what you see in the recording above). What we could do is go for option b) for now and swap for option a) once mobile Safari supports `preventScroll`.

**Comment by jasmussen** [comment]
Absolutely amazing work, Michael. You should feel free to push your changes to this branch. It looks like a tiny hack that's very isolated, and our understanding of why it's necessary and how it can be one day deprecated (once Mobile Safari improves) is good. Combined with how VASTLY better this is compared to what's in `master`, I say ship it!

I'll still need to remove some of the hacky experiments I wrote in this branch, and to unhide the adminbare again (perhaps explore sticky positioning again). But with that done, and your iOS changes, this should be good for a re-review.

**Comment by frontdevde** [comment]
Encouraged by @jasmussen I pushed a somewhat suboptimal fix. There's probably a better solution out there and I'm continuing to look into it. I pushed it up though as for now, this fix does solve two of the main issues though that we're seeing: On the one hand, the issue of the navigation being scrolled off-screen by the browser on focus. On the other hand, once a block is focussed it also appears to remove any whitespace that might have been introduced by the overscroll behavior on mobile Safari.

**Comment by frontdevde** [comment]
I just tested this PR with the navigation block and appears to fix the issue of the navigation block breaking on mobile due to submenu items' toolbar (see https://github.com/WordPress/gutenberg/issues/18401).

ezgif com-video-to-gif (3)

File: packages/block-editor/src/components/block-mobile-toolbar/index.js [hunk 3]
```diff
@@ -8,6 +8,19 @@ import { ifViewportMatches } from '@wordpress/viewport';
...
+// Once supported in Safari we can replace this in favor of preventScroll.
+// For details see issue #18632 and PR #18686
+const isIphone = window.navigator.userAgent.indexOf( 'iPhone' ) !== -1;
+if ( isIphone ) {
+	document.addEventListener( 'focusin', function( ) {
+		setTimeout( () => {
+			window.scrollTo( 0, 0 );
+		}, 150 );
+	} );
+}
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-mobile-toolbar/index.js]: While I don't understand the hack entirely, I do feel like maybe it's not the right place for it and maybe something for the edit-post package instead (like in the index or a hack imported there)

Comment (jasmussen) [reply packages/block-editor/src/components/block-mobile-toolbar/index.js]: I'll try to explain the hack to the best of my ability. We might even want to include this as a comment in context of the hack:

- On mobile,`edit-post-editor-regions__body` scrolls to enable a fixed top toolbar.
- On an iOS device, when the mobile keyboard pops up, Mobile Safari forces `html` to scroll upwards, even when it shouldn't, moving the top toolbar out of view.
- This compensates for that onfocus scrolling.

**Comment by jasmussen** [comment]
Gave it a good rebase and squash to simplify and keep the branch fresh. I'm going to look into some of the hack work I've done, see how much I can remove, and to resurface the wp-admin toolbar again.

**Comment by jasmussen** [comment]
Removed some hacks, but kept one in that was necessary now that the adminbar is showing again. Also added some additional comments around the iOS hack. 

The scroll hack looks slightly more pronounced now due to the adminbar showing:

scrolling

— there's more to scroll! This is another reason to look at refactoring that away.

I actually tried including `scroll-behavior: smooth;` on the `html` element to see if that would somehow make the scroll hack less jarring. Alas, it did not.

**Comment by jasmussen** [comment]
One last thing worth either fixing or looking into, it's most visible in this screenshot of just a small viewport and with scrollbars forcibly enabled:

Screenshot 2019-11-27 at 13 09 33

The top formatting toolbar actually _covers content_. Are we okay with that? I am — I don't think it's the end of the world — but worth thinking about. 

Two alternative fixes:

1. Always have permanent white visible space (white empty bar with gray border) where the toolbar will be. 
2. Add top padding the height of the toolbar, or even topmargin, when the toolbar is showing. This will cause a shift when it appears.

**Comment by karmatosed** [comment]
I came here to just leave a comment about being very excited of all the great work has gone on so far. It's a tough thing to solve and already feels so much better looking in screenshots.

**Comment by jasmussen** [comment]
Unless there's disagreement about what to do in https://github.com/WordPress/gutenberg/pull/18686#issuecomment-559062014, I think we should probably get this merged as it's a step forward for iOS users. 

I'll rebase and see if the checks pass.

**Comment by jasmussen** [comment]
Actually, this rebase is gnarly. @youknowriad if you have a moment, can you look?

**Comment by youknowriad** [comment]
this still need to be moved to the right place https://github.com/WordPress/gutenberg/pull/18686#discussion_r351206241 (I can do it a bit later if needed)

**Comment by youknowriad** [comment]
Thanks for the updated @frontdevde Feel free to rebase and land :)

**Comment by frontdevde** [comment]
@jasmussen I just rebased this branch with the current master. There were quite a few conflicts that had to be resolved. Could you help test if everything is still working as expected?

**Comment by jasmussen** [comment]
Current status:

status

Yep, this feels very good to land. I think it's good to go. I also tested fullscreen mode and desktop viewpoints and all looks good.

There's one issue we want to look at, but this is a separate issue likely related to moving the mover control to the toolbar, so it could be worth looking at that then (cc @youknowriad as I think you've explored that). Or it needs to happen in the popover component itself:

Screenshot 2019-12-04 at 17 52 26

But again, not a blocker for this PR. I say ship it!

**Comment by mtias** [comment]
Thank you all for working on this!

**Comment by hypest** [comment]
It seems that a style filename change in this PR broke the native mobile codepath because [there's an import](https://github.com/WordPress/gutenberg/blob/391c225aab2b0f6eb11afd66a05fafa4b4126477/packages/block-editor/src/components/block-mobile-toolbar/index.native.js#L17) that wasn't renamed to match the new filename. [A PR is already up to fix this](https://github.com/WordPress/gutenberg/pull/18935) but I'd like to take the opportunity to raise awareness to the React Native checkbox in the [the PR template checklist](https://github.com/WordPress/gutenberg/blob/master/.github/PULL_REQUEST_TEMPLATE.md). The checklist seems to be missing from this PR. Thanks!
